### PR TITLE
Get schema json asynchronously

### PR DESCRIPTION
**Pull request by AkshayN5**
[Corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/263015)

### CODE REVIEW

**Comment by pmconne** [comment]
Do you intend to deprecate `getSchemaSync` to discourage its usage?

**Comment by AkshayN5** [comment]
> Do you intend to deprecate `getSchemaSync` to discourage its usage?

yes, and we also want to do some performance testing with the async method to make sure that it helps optimize for memory usage and runtime.

File: core/backend/src/IModelSchemaLoader.ts [hunk 1]
```diff
@@ -145,22 +144,15 @@ class IModelSchemaLocater implements ISchemaLocaterType {
...
-    const val = await new Promise<IModelJsNative.ErrorStatusOrResult<any, string>>((resolve) => {
-      this._iModel.nativeDb.getSchemaAsync(schemaName, resolve);
-    });
-
-    if (undefined !== val.error) {
-      if (IModelStatus.NotFound === val.error.status) {
-        return undefined;
-      }
-      throw new IModelError(val.error.status, `reading schema=${schemaName}`);
+    try {
```
Comment (kabentley) [inline core/backend/src/IModelSchemaLoader.ts:152]: why not just let this throw an error so the caller has a chance of figuring out what's wrong?

Comment (kabentley) [reply core/backend/src/IModelSchemaLoader.ts:159]: There's no point in catching an exception and then rethrowing it. This method is useless.

File: core/backend/src/IModelDb.ts [hunk 2]
```diff
@@ -1108,6 +1108,15 @@ export abstract class IModelDb extends IModel {
     return this.nativeDb.getMassProperties(JsonUtils.toObject(props));
   }
 
+  /** Get the serialized schema in JSON format of the given name asynchronously. */
+  public async getSchemaAsync(schemaName: string): Promise<string> {
+    try {
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:1113]: remove try/catch

**Comment by kabentley** [comment]
you probably need to run `rush extract-api` and `rush change`

**Comment by kabentley** [comment]
i'm not a reviewer or I would approve

**Comment by pmconne** [comment]
> you probably need to run `rush extract-api` and `rush change`

@AkshayN5 in case you have not read [addon development guidelines](https://dev.azure.com/bentleycs/iModelTechnologies/_wiki/wikis/iModelTechnologies.wiki/8690/iModel.js-Node-add-on-development).

**Comment by AkshayN5** [comment]
> > you probably need to run `rush extract-api` and `rush change`
> 
> @AkshayN5 in case you have not read [addon development guidelines](https://dev.azure.com/bentleycs/iModelTechnologies/_wiki/wikis/iModelTechnologies.wiki/8690/iModel.js-Node-add-on-development).



> > you probably need to run `rush extract-api` and `rush change`
> 
> @AkshayN5 in case you have not read [addon development guidelines](https://dev.azure.com/bentleycs/iModelTechnologies/_wiki/wikis/iModelTechnologies.wiki/8690/iModel.js-Node-add-on-development).

I run `rush extract-api` but it did not update any signature files in the `common/api` directory

**Comment by pmconne** [comment]
> I run `rush extract-api` but it did not update any signature files in the `common/api` directory

Did you build first? You added a new public method called `getSchemaAsync` that I would expect to be found by extract-api.
Please do:
```
rush build --to core-backend
cd core/backend
npm run extract-api
```

**Comment by AkshayN5** [comment]
> Did you build first? You added a new public method called `getSchemaAsync` that I would expect to be found by extract-api. Please do:
> 
> ```
> rush build --to core-backend
> cd core/backend
> npm run extract-api
> ```

I tried running these commands. However, it still did not update any signature files in the common/api directory

**Comment by pmconne** [comment]
> I tried running these commands. However, it still did not update any signature files in the common/api directory

Ah, the class is not exported.

**Review by pmconne — APPROVED** [review]

**Comment by pmconne** [comment]
@AkshayN5 what's the status of this?

**Comment by kabentley** [comment]
@AkshayN5 @ColinKerr How can this PR possibly take a **month** to complete? It will now conflict with my other PR. Can **someone** please do something about this?

**Comment by AkshayN5** [comment]
> @AkshayN5 @ColinKerr How can this PR possibly take a **month** to complete? It will now conflict with my other PR. Can **someone** please do something about this?

This PR is currently on hold as the performance of the newly added async methods is slow as compared to the existing sync method.

**Comment by calebmshafer** [comment]
converted this to a draft PR due to the approach not being ready for review and still needs dev work.

**Comment by ColinKerr** [comment]
> > @AkshayN5 @ColinKerr How can this PR possibly take a **month** to complete? It will now conflict with my other PR. Can **someone** please do something about this?
> 
> This PR is currently on hold as the performance of the newly added async methods is slow as compared to the existing sync method.

To put some numbers to this small schemas were ~100x (single digit ms to 100s of ms) slower and larger schemas were ~5x slower (100-200ms to 400-600ms).  So we need to understand why before we move forwards but bug fixes and more in demand performance enhancements now have greater priority on our backlog.

**Comment by ColinKerr** [comment]
> > > @AkshayN5 @ColinKerr How can this PR possibly take a **month** to complete? It will now conflict with my other PR. Can **someone** please do something about this?
> > 
> > 
> > This PR is currently on hold as the performance of the newly added async methods is slow as compared to the existing sync method.
> 
> To put some numbers to this small schemas were ~100x (single digit ms to 100s of ms) slower and larger schemas were ~5x slower (100-200ms to 400-600ms). So we need to understand why before we move forwards but bug fixes and more in demand performance enhancements now have greater priority on our backlog.

I revisited the async tests and found an issue with a dangling promise so the test runs were overlapping.  I missed this when I first reviewed the tests, my bad.  With this fixed the async code is 1.1x-2.3x slower with an average of 1.5x slower.  This is much more reasonable, so I will convert this back to a full PR.

**Comment by ColinKerr** [comment]
> Corresponding native PR

[New PR](https://bentleycs.visualstudio.com/iModelTechnologies/_git/imodel02/pullrequest/271724)

File: full-stack-tests/backend/src/perftest/SchemaLoader.test.ts [hunk 3]
```diff
@@ -4,12 +4,12 @@
...
 import * as path from "path";
 import { Reporter } from "@itwin/perf-tools";
-import { IModelJsFs, IModelJsNative, KnownLocations, StandaloneDb } from "@itwin/core-backend";
+import { IModelHost, IModelJsFs, KnownLocations, StandaloneDb } from "@itwin/core-backend";
 import { IModelTestUtils, KnownTestLocations } from "@itwin/core-backend/lib/cjs/test";
 import * as fs from "fs";
 import { OpenMode } from "@itwin/core-bentley";
 
-describe("SchemaLoaderPerformance", () => {
+describe.only("SchemaLoaderPerformance", () => {
```
Comment (pmconne) [inline full-stack-tests/backend/src/perftest/SchemaLoader.test.ts:12]: remove `.only`.

Comment (ColinKerr) [reply full-stack-tests/backend/src/perftest/SchemaLoader.test.ts]: Done, also removed errant console log.

**Review by ColinKerr — APPROVED** [review]

**Pull request merged**